### PR TITLE
feat(quest-details): add support for individual claim click

### PR DESCRIPTION
**Pull request by eliobricenov**
### Summary

Overall the change is to replace the global claim all logic to individual claim actions for each reward. [Story Link](https://quests-ui-git-feat-claim-quest-page-hyperplay.vercel.app/?path=/story/components-questdetailswrapper--quest-page-not-signed-in)

### CODE REVIEW

File: src/state/QuestWrapperProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  syncPlaySession: (appName: string, runner: Runner) => Promise<void>
+  logInfo: (message: string) => void
+  openDiscordLink: () => void
+  getDepositContracts: (questId: number) => Promise<DepositContract[]>
+  tOverride?: TFunction<any, string>
+  sessionEmail?: string
+  checkG7ConnectionStatus: () => Promise<boolean>
+  isQuestsPage?: boolean
+  onPlayClick: (quest: Quest) => void
+  onRewardsClaimed?: (rewards: Reward[]) => void
```
Comment (BrettCleary) [inline src/state/QuestWrapperProvider.tsx:47]: this looks to be used only for the quest toast that is shown, first setting the claimed rewards in the `useClaimedRewards` hook https://github.com/HyperPlay-Gaming/hyperplay-desktop-client/pull/1132/files#diff-5005a854a0db0c122c6e1d2fc7ff9c3569aea702bbcb61f5fc915ad1fb781bc7

should we move that hook's state to a separate mobx class that acts as a toast/notification system generally or just for quests? That way all future quest notifications are handled by that mobx class so long as it has been initialized by the consuming code without having to add a new hook and pass a new prop here

File: src/components/RewardWrapper/index.tsx [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+import { useTranslation } from 'react-i18next'
+import { createPublicClient, http } from 'viem'
+import { useAccount, useConnect, useWriteContract, useSwitchChain } from 'wagmi'
+import { injected } from 'wagmi/connectors'
+import { ConfirmClaimModal } from '../ConfirmClaimModal'
+import { AlertCard, Reward as RewardUi } from '@hyperplay/ui'
+import styles from './index.module.scss'
+import { UseGetRewardsData } from '@/types/quests'
+
+class ClaimError extends Error {
```
Comment (BrettCleary) [inline src/components/RewardWrapper/index.tsx:25]: could move this to `types/quests.ts` for future reuse

File: src/helpers/isEligibleForClaiming.ts
Comment (BrettCleary) [inline src/helpers/isEligibleForClaiming.ts:1]: can you use `getPlaystreakQuestStatus` at `src\helpers\getPlaystreakQuestStatus.ts` in this repo? It has tests and handles some edge cases that this helper function does not

File: src/state/QuestWrapperProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+} from '@hyperplay/utils'
+
+interface QuestWrapperContextValue {
+  flags: {
+    rewardTypeClaimEnabled: Record<Reward['reward_type'], boolean>
+    questsOverlayClaimCtaEnabled?: boolean
+  }
+  getQuest: (questId: number) => any
+  getUserPlayStreak: (questId: number) => any
+  getSteamGameMetadata: (id: number) => any
```
Comment (BrettCleary) [inline src/state/QuestWrapperProvider.tsx:20]: these should be return promises right?

File: src/hooks/useGetRewards.ts [hunk 4]
```diff
@@ -3,14 +3,19 @@ import { useGetQuest } from './useGetQuest'
...
-  getExternalTaskCredits: (rewardId: string) => Promise<string>,
+export function useGetRewards({
+  questId,
+  getQuest,
+  getExternalTaskCredits,
+  logError
+}: {
+  questId: number | null
+  getQuest: (questId: number) => Promise<Quest>
+  getExternalTaskCredits: (rewardId: string) => Promise<string>
```
Comment (BrettCleary) [inline src/hooks/useGetRewards.ts:16]: Could get these types from `QuestWrapperContextValue` to prevent redefining them.

File: src/state/QuestWrapperProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import {
+  ConfirmClaimParams,
+  DepositContract,
+  Quest,
+  Reward,
+  RewardClaimSignature,
+  Runner
+} from '@hyperplay/utils'
+
+interface QuestWrapperContextValue {
```
Comment (BrettCleary) [inline src/state/QuestWrapperProvider.tsx:13]: Could we move this to `types/quests`? That way functions like your `useGetRewards` hook could import it and use its types more easily

File: package.json
Comment (BrettCleary) [inline package.json:1]: Should we go ahead and add each of the new wrapper components as exports in `src\index.ts`?

File: src/components/RewardsWrapper/index.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+              questPlayStreakData={questPlayStreakData?.userPlayStreak}
+              invalidateQuestPlayStreakQuery={invalidateQuestPlayStreakQuery}
+            />
+          ))}
+        </RewardsRow>
+      </Fragment>
+    )
+  )
+
+  return <Rewards>{rewardsContent}</Rewards>
```
Comment (BrettCleary) [inline src/components/RewardsWrapper/index.tsx:66]: should we return a loading spinner here while waiting on use get rewards to return?

File: src/components/RewardsWrapper/index.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  // Organize rewards by category
+  const rewardsByCategory: Record<string, UseGetRewardsData[]> = {}
+
+  for (const reward_i of rewardsData) {
+    if (Object.hasOwn(rewardsByCategory, reward_i.chainName)) {
+      rewardsByCategory[reward_i.chainName].push(reward_i)
+    } else {
+      rewardsByCategory[reward_i.chainName] = [reward_i]
+    }
+  }
```
Comment (BrettCleary) [inline src/components/RewardsWrapper/index.tsx:45]: wondering if we should make a helper for this and use it/return `rewardsByCategory` from `useGetRewards`. It's O(n) so probably fine to just run it every time `useGetRewards` runs even if the consuming code doesn't necessarily need the map.

File: src/components/QuestDetailsWrapper/index.tsx
Comment (BrettCleary) [inline src/components/QuestDetailsWrapper/index.tsx:1]: 750 to 300 lines. That's awesome! Now the logic that was in here is in the respective subcomponent. This is a great and much needed refactor 🔥 🔥 🔥

File: src/components/QuestDetailsWrapper/PlayStreakQuest.stories.tsx
Comment (BrettCleary) [inline src/components/QuestDetailsWrapper/PlayStreakQuest.stories.tsx:1]: fantastic job on these stories 🚀

**Review by BrettCleary — CHANGES_REQUESTED** [review]
this pr is a beast! requesting changes mainly for the `isEligibileForClaiming` helper. also left some small questions and comments

File: package.json
Comment (BrettCleary) [reply package.json:1]: sure we can add the exports on a future PR when they are needed

**Comment by BrettCleary** [comment]
looks like there are a few conflicts now @eliobricenov

File: package.json
Comment (BrettCleary) [inline package.json:1]: can you bump the version here by a minor version?

**Review by BrettCleary — APPROVED** [review]
LGTM. thx for making the changes. let's ship it and test e2e in the client on your claim branch

File: src/state/ClaimedRewardToastState.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+class ClaimedRewardToastState {
+  private reward: Reward | null = null
+
+  constructor() {
+    makeAutoObservable(this)
+  }
+
+  showClaimedReward(reward: Reward) {
+    this.reward = reward
+  }
```
Comment (BrettCleary) [inline src/state/ClaimedRewardToastState.ts:13]: can we add a 5 second timeout here to clear the reward? and if it is claimed while this timer is active, it resets back to 5 more seconds to timeout

**Review by BrettCleary — APPROVED** [review]
lgtm!
